devito/arguments.py

```diff
@@ -1,152 +1,133 @@
+
+The various class hierarchies are explained here:
+Parameter: Any node of the dependency graph has to necessarily be of this type. The node
+           may or may not represent an actual runtime argument passed to the kernel.
+Argument: Subclass of Parameter. This represents a node in the dependency graph directly
+          corresponding to a runtime argument passed to the kernel.
+
+             Parameter
+                 |
+        -------------------
+DimensionParameter        |
+                          |
+                       Argument
+                          |
+                          |
+                --------------------
+                |         |        |
+        ScalarArgument    |        |
+                    TensorArgument |
+                                PtrArgument
+
+ArgumentEngine: The main external API for this module. It encapsulates all the argument
+                derivation and verification logic.
+
+ArgumentVisitor: Visits objects of devito's data types to return appropriate objects
+                 from the above parameter-argument hierarchy.
+
+ValueVisitor: Used by the argument derivation method to derive the value of each parameter
+              based on the dependency tree.
+
+Dependency: Edges of the dependency graph are represented by objects of this class.
+
+UnevaluatedDependency: A "future" object representing an argument derivation that is yet
+                       to happen.
 """
 
 
-class Argument(object):
+class Parameter(object):
+    """ Abstract base class for any object that represents a node in the dependency
+        graph. It may or may not represent a runtime argument.
 
-    """ Abstract base class for any object that represents a run time argument for
-        generated kernels.
+    :param name: Name of the parameter
+    :param dependencies: A list of :class:`Dependency` objects that represent all the
+                         incoming edges into this node of the graph.
     """
-
-    __metaclass__ = abc.ABCMeta
-
+    is_Argument = False
     is_ScalarArgument = False
     is_TensorArgument = False
     is_PtrArgument = False
 
-    def __init__(self, name, provider, default_value=None):
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, name, dependencies):
         self.name = name
-        self.provider = provider
-        self.default_value = default_value
-        self.reset()
+        self.dependencies = dependencies
 
     @property
-    def value(self):
-        try:
-            if self._value.is_SymbolicFunction:
-                return self._value._data_buffer
-            elif self._value.is_Constant:
-                return self._value.data
-            else:
-                raise InvalidArgument("Unexpected data object %s" % type(self._value))
-        except AttributeError:
-            return self._value
+    def gets_value_from(self):
+        return [x for x in self.dependencies if x.dependency_type == "gets_value_from"]
```

Comment:
``gets_value_from`` is still a magic string

Reply:
Fixed

Reply:
Pushed now.